apps/email/js/cards/cmp/autocomplete.js

```diff
@@ -0,0 +1,476 @@
+        while ((node = node.parentNode)) {
+          if (node.classList && node.classList
+                                .contains('cmp-autocomplete-item')) {
+            break;
+          }
+        }
+      }
+
+      if (node && node.match) {
+        this.dispatchEvent(new CustomEvent('autocompleteSelected', {
+          detail: {
+            match: node.match,
+            inputNode: this.inputNode
+          }
+        }));
+        this._matchClickedTime = Date.now();
+      }
+    },
+
+    /**
+     * Loads the dataSource module.
+     * @param  {String} dataSourceId The module ID for the data source.
+     */
+    bindDataSource: function(dataSourceId) {
+      require([dataSourceId], (dataSource) => {
+        this.dataSource = dataSource;
+        if (this.query) {
+          this.match();
+        }
+      });
+    },
+
+    /**
+     * Called to trigger finding matches from the dataSource. Set on a timeout
+     * to avoid multiple fast keystrokes from triggering too many rapid queries.
+     */
+    match: function() {
+      if (!this.query || !this.dataSource) {
+        this.list.innerHTML = '';
+        this.hideList();
+        return;
+      }
+
+      if (!this.timedQueryId) {
+        // The choice of delay tries to balance sending too many queries to the
+        // contacts API vs giving a feeling of responsiveness for the reaction
+        // of the autocomplete. This balance may be affected by the total number
+        // of contacts.
+        this.timedQueryId = setTimeout(this.runQuery, 350);
+      }
+    },
+
+
+    /**
+     * The owner of this autocomplete instance should set this value if they
+     * have extra spacing around the input that should be maintained for better
+     * visual display, and to avoid small scroll janks on initial positioning
+     * of the input element if it needs to be scrolled to the top of the view.
+     */
+    verticalSpace: 0,
+
+    /**
+     * This method can be overriden by the owner of the autocomplete to return
+     * existing entries for the given input node. The result should be an
+     * array of objects that have "name" and "address" properties.
+     * @param  {Element} inputNode The input element that is receiving the
+     * autocomplete.
+     * @return {Array}   Array of existing entries.
+     */
+    getExistingEntries: function(inputNode) {
+      return [];
+    },
+
+    /**
+     * Runs the query to find matches from the dataSource.
+     */
+    runQuery: function() {
+      this.timedQueryId = 0;
+
+      this.dataSource(this.query).then((result) => {
+        // Protect against old or out of date queries
+        if (result.query !== this.query) {
+          return;
+        }
+
+        var existingEntries = this.getExistingEntries(this.inputNode);
+
+        var contacts = result.contacts;
+
+        this.list.innerHTML = '';
+        this.topMatch = null;
+
+        var queryRegExp = new RegExp(this.query, 'i');
```

Comment:
Thanks for the catch, I had a mental note to look into it, but forgot to write it down. I created a regExpEscape module that just returns a function, using the polyfill code:

```javascript
'use strict';
/**
 * Taken from the ECMAScript proposal polyfill in the repo:
 * https://github.com/benjamingr/RegExp.escape
 * git commit: de7b32f9f72d15f9d580119963139d1027c46bbe
 * Copyright CC0 1.0 Universal Benjamin Gruenbaum 2015
 */

define(function() {
  return function regExpEscape(str) {
    return str.replace(/[\\^$*+?.()|[\]{}]/g, '\\$&');
  };
});
```